src/main/java/hudson/tasks/junit/SuiteResult.java

```diff
@@ -235,14 +299,36 @@ private SuiteResult(File xmlReport, Element suite, boolean keepLongStdio) throws
 
     /*package*/ void addCase(CaseResult cr) {
         cases.add(cr);
-        casesByName().put(cr.getName(), cr);
-        
+        casesByName().put(cr.getDisplayName(), cr);
+
         //if suite time was not specified use sum of the cases' times
         if( !hasTimeAttr() ){
             duration += cr.getDuration();
         }
     }
 
+    @Nonnull
+    private String getEnclosingNodeDisplayName(@Nonnull WorkflowRun run, @Nonnull String nodeId) {
+        FlowExecution execution = run.getExecution();
+        if (execution != null) {
+            try {
+                FlowNode node = execution.getNode(nodeId);
+                if (node != null) {
+                    if (node.getAction(LabelAction.class) != null) {
+                        ThreadNameAction threadNameAction = node.getAction(ThreadNameAction.class);
+                        if (threadNameAction != null) {
+                            return threadNameAction.getThreadName();
+                        }
+                    }
+                    return node.getDisplayName();
+                }
+            } catch (Exception e) {
+                // TODO: Something in that odd case where we can get an NPE
```

Comment:
Wait, NPE where?  It looks like this is null protected?

Reply:
`execution.getNode` can theoretically throw an NPE. =)